File: Patterns/Registration/Python/TestPackage/Base/ComponentBuilder.py
# Import Top level subpackages
from TestPackage import SubPackage_01
from TestPackage import SubPackage_02

# Import Registration
from .ComponentRegistration import ComponentRegistration as register

class ComponentBuilder():

    @staticmethod
    def build(name):
        assert name in register.componentMap.keys(), "Unknown Component: {0:s}. Must be on of [{1:s}]".format(name, "|".join(register.componentMap.keys()))
        return register.componentMap[name]()

File: Patterns/Registration/Cpp/TestPackage/Base/ComponentBuilder.cpp
#include "ComponentBuilder.h"
#include "ComponentRegistration.h"
#include <sstream>
#include <vector>
#include <iterator>

std::shared_ptr<ComponentBase> ComponentBuilder::build(const std::string &name){
    auto registry = ComponentRegistration::getRegistry();
    // Check that requested component is in registry
    if (registry.find(name) == registry.end()){
        // Gather Keys
        std::vector<std::string> keys;
        for (const auto item : registry)
            keys.push_back(item.first);
        // Write Error message
        std::stringstream strm;
        strm << "Unknown Component: " << name << ". Must be one of [";
        std::copy(keys.begin(), keys.end(), std::ostream_iterator<std::string>(strm, "|"));
        strm << "]";
        throw std::runtime_error(strm.str());
    }

    return registry[name]();
}